bokehjs/src/coffee/models/scales/scale.coffee

```diff
@@ -2,22 +2,29 @@ import {Transform} from "../transforms"
 import * as p from "core/properties"
 
 export class Scale extends Transform
+  type: "Scale"
 
   @internal {
     source_range: [ p.Any ]
-    target_range: [ p.Any ]
+    target_range: [ p.Any ] # p.Instance(Range1d)
   }
 
-  # map to/from methods are deprecated (provided for back compat)
+  compute: (x) ->
 
-  map_to_target: (x) ->
-    @compute(x)
+  v_compute: (xs) ->
 
-  v_map_to_target: (xs) ->
-    @v_compute(xs)
+  invert: (sx) ->
 
-  map_from_target: (xprime) ->
-    @invert(xprime)
+  v_invert: (sxs) ->
 
-  v_map_from_target: (xprimes) ->
-    @v_invert(xprimes)
+  r_compute: (x0, x1) ->
+    if @target_range.is_reversed
+      return [@compute(x1), @compute(x0)]
+    else
+      return [@compute(x0), @compute(x1)]
+
+  r_invert: (sx0, sx1) ->
+    if @target_range.is_reversed
+      return [@invert(sx1), @invert(sx0)]
+    else
+      return [@invert(sx0), @invert(sx1)]
```

Comment:
I don't really like these functions. It's potentially very confusing to have the order of the parameters coupled implicitly in a hidden way to a property of some other model. It also muddies the conceptual clarity and simplicity of "compute" which really should just be "compute the transform of this input". The mapping of `x0` is what it is, it should not be conflated with a mapping for a completely different value in some implicit way.

Reply:
I also don't understand why this would be needed? If this is just to try and keep the "order" or two arbitrary inputs in the same order as a range, we already have "bbox" function to make sure start < end, etc. that are used where appropriate.

Reply:
I think it would be very confusing if I implemented `v_invert/v_compute` like this. This is why I introduced `r_` variant for pairs of coordinates. `r_` variant simply maintains spatial order. If you don't want this, then you use `v_` variant or multiple calls to `invert/compute`. I tried multiple solutions so far, and this one is the cleanest and allows for very straightforward and explicit maintenance of transition between data and screen space ranges. Previously there were multiple ways to achieve spatial order, now there is one. No implicit conversions are necessary, no variable swapping, no reversed inequalities. I presume this change may make scales less pure, but it makes everything else much easier to work with.

Reply:
I was definitely not suggesting `v_invert/v_compute` be implemented this way. I was suggesting not having this at all, and making callers explicitly responsible for caring (or not) about order. 

Because this does not prevent misuse. What if I call `r_compute(x1, x0)` then then values get swapped when they shouldn't be. And it doesn't help at all if I need to convert more than two values for some reason. Isn't the actual intent of this function simply that the  results be sorted in the same order as the target range? If so, I'd propose adding a `sort_by_target` flag to `v_compute` and does nothing if false (current behavior), or sorts ascending or descending based on `is_reversed` if true. That has semantics that don't force an artificial start/end concept on input data that does not have it, works correctly regardless of what order args are passed in, and has semantics that make more sense (to me).

Reply:
Alternatively `sorted_compute`, `v_sorted_compute` but personally I'd prefer the flag

Reply:
I guess my takeaway is: actually do a real sort instead of the fake-sort-by-swapping. Because the "swapping" only works coincidentally, under a specific subset of circumstances, whereas doing a real sort will always work and also make the semantics consistent and sensible.

Reply:
>  but personally I'd prefer the flag

Abstracting from this PR, adding a flag is almost always a very bad idea, and a terrible one in JS, where you don't have named arguments. The two variations must be easily distinguishable and searchable, so there must be different names. Internally one can be implemented in terms of the other.

Reply:
I agree JS makes it worse, so as I say I'd be fine with `sorted_compute, v_sorted_compute` then

Reply:
There is no point in having `v_` sorted variant, because the only usage of the existing `r_` variant is for pairs of coordinates. That's why this function takes two scalar arguments and returns a 2-tuple. Previously this was emulated with `v_compute([arg0, arg1])`, but that's not semantically equivalent, and it will be better visible when I add typings for scales.

Reply:
> I guess my takeaway is: actually do a real sort instead of the fake-sort-by-swapping.

I disagree that what's needed here is sorting. The goal of `r_invert` is to maintain the order of output coordinates the same as the order of inputs. However, you may be right, but I need to give this some more thought and look at this from a fresh perspective.

Reply:
The output of this function is highly dependent on the order of the arguments, but there is indication in the function name that this is the case, or further, what that order needs to be. If I have two parameters `xfoo` and `xbar` there is nothing to tell me whether to call `func(xfoo, xbar)` or to call `func(xbar, xfoo)`. The only way I can be sure to call things correctly is to know to know something about the function (that it might swap the order) and to then put the args in the right order by inspecting them and taking that into account. If you ask me to say what this function *does* I can't say, swapping unconditionally based on the range is nonsensical in many situations that seem easy to be in.